stream-legal-entity/legal-entity-core/src/main/java/com/backbase/stream/LegalEntitySaga.java

```diff
@@ -159,9 +143,113 @@ public Mono<LegalEntityTask> executeTask(@SpanTag(value = "streamTask") LegalEnt
+        if (externalUserId == null) {
+            streamTask.info(LEGAL_ENTITY, PROCESS_CONTACTS, FAILED, legalEntity.getExternalId(), legalEntity.getInternalId(),
+                    "Legal Entity: %s does not have any Users", legalEntity.getExternalId());
+            return Mono.just(streamTask);
+        }
+        return contactsSaga.executeTask(createContactsTask(streamTask.getId(), legalEntity.getExternalId(), serviceAgreement.getExternalId(), externalUserId, AccessContextScope.LE, legalEntity.getContacts()))
+                .flatMap(contactsTask -> requireNonNull(Mono.just(streamTask)))
+                .then(Mono.just(streamTask));
+    }
+
+    private Mono<LegalEntityTask> postServiceAgreementContacts(LegalEntityTask streamTask) {
+        LegalEntity legalEntity = streamTask.getData();
+        ServiceAgreement serviceAgreement = getServiceAgreement(legalEntity);
+        if (isEmpty(serviceAgreement.getContacts())) {
+            streamTask.info(SERVICE_AGREEMENT, PROCESS_CONTACTS, FAILED, serviceAgreement.getExternalId(), serviceAgreement.getInternalId(),
+                    "Master Service Agreement: %s does not have any Contacts defined", serviceAgreement.getExternalId());
+            return Mono.just(streamTask);
+        }
+        log.info("Creating Contacts for Service Agreement Id {}", serviceAgreement.getExternalId());
+        String externalUserId = getUserExternalId(legalEntity.getUsers());
+        if (externalUserId == null) {
+            externalUserId = getParticipantUser(serviceAgreement);
+            if (externalUserId == null) {
+                streamTask.info(LEGAL_ENTITY, PROCESS_CONTACTS, FAILED, legalEntity.getExternalId(), legalEntity.getInternalId(),
+                        "Legal Entity: %s does not have any Users", legalEntity.getExternalId());
+                return Mono.just(streamTask);
+            }
+        }
+        return contactsSaga.executeTask(createContactsTask(streamTask.getId(), legalEntity.getExternalId(), serviceAgreement.getExternalId(), externalUserId, AccessContextScope.SA, serviceAgreement.getContacts()))
+                .flatMap(contactsTask -> requireNonNull(Mono.just(streamTask)))
+                .then(Mono.just(streamTask));
+    }
+
+    private String getParticipantUser(ServiceAgreement serviceAgreement) {
+        if (!isEmpty(serviceAgreement.getParticipants())) {
+            Optional<LegalEntityParticipant> participants = serviceAgreement.getParticipants()
+                    .stream()
+                    .filter(LegalEntityParticipant::getSharingUsers)
+                    .findFirst();
+            if (participants.isPresent()) {
+                LegalEntityParticipant participant = participants.get();
+                if (!isEmpty(participant.getUsers())) {
+                    return getOptionalUserId(participant.getUsers().stream().findFirst());
+                } else if (!isEmpty(participant.getAdmins())) {
```

Comment:
shouldn't admins be highest in the preference order?

Reply:
changed to admin first